### PR TITLE
New grok pattern input component

### PR DESCRIPTION
**Pull request by kmerz**
## Description
Create a new grok input component which will help a user creating a new
grok input from existing ones.

A user can search via filter for grok patterns and add them to the
text field beside him. A grok pattern will be placed under the current
cursor position of the text field.

## Motivation and Context
If you want to create a new grok pattern from existing ones you need to open a second
tap so you can search for the existing ones. To prevent such behavior we now have
the grok pattern filter which can also add the grok patterns to the text field.

## How Has This Been Tested?
Editing or creating a grok pattern and use the filter function to find a grok pattern and add it with 'Add' button.

## Screenshots (if appropriate):
screenshot-2018-5-28 graylog - grok patterns

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by bernd** [comment]
Any progress on this? :smiley:

File: graylog2-web-interface/src/components/bootstrap/BootstrapModalWrapper.jsx [hunk 1]
```diff
@@ -42,7 +42,7 @@ class BootstrapModalWrapper extends React.Component {
 
   render() {
     return (
-      <Modal show={this.state.showModal} onHide={this.hide}>
+      <Modal show={this.state.showModal} onHide={this.hide} bsSize="large">
```
Comment (edmundoa) [inline graylog2-web-interface/src/components/bootstrap/BootstrapModalWrapper.jsx]: We should probably add a prop to control that `bsSize`, and I would also add it to `BootstrapModalForm`. In that way we can customize this when we need to.

File: graylog2-web-interface/src/components/grok-patterns/GrokPatterns.jsx [hunk 2]
```diff
@@ -97,6 +97,7 @@ const GrokPatterns = createReactClass({
                               name={pattern.name}
                               pattern={pattern.pattern}
                               testPattern={this.testPattern}
+                              patterns={this.state.patterns}
```
Comment (edmundoa) [inline graylog2-web-interface/src/components/grok-patterns/GrokPatterns.jsx]: We should filter the current pattern being edited by the user, as that should not be added to the Grok pattern.

File: graylog2-web-interface/src/components/grok-patterns/GrokPatternInput.jsx [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        activeListItem = this.state.activeListItem - 1;
+        if (activeListItem < 0) {
+          return;
+        }
+        domElement = document.getElementById(`list-item-${activeListItem}`);
+        domElement.scrollIntoView();
+        this.setState({ activeListItem: activeListItem });
+        e.preventDefault();
+        break;
+      case ENTER:
```
Comment (edmundoa) [inline graylog2-web-interface/src/components/grok-patterns/GrokPatternInput.jsx]: This does not properly work when pressing enter while the focus is in the filter, example:

- Go to the input filter
- Type anything (optional)
- Press enter
- `%{undefined}` is added to the list

File: graylog2-web-interface/src/components/grok-patterns/GrokPatternInput.jsx [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        const active = index === this.state.activeListItem;
+        this.shownListItems.push(pattern.name);
+        return (
+          <ListGroupItem id={`list-item-${index}`}
+                         header={pattern.name}
+                         bsStyle={active ? 'info' : undefined}
+                         onKeyDown={this._onPatternFilterKeyDown}
+                         key={pattern.name}>
+            <span>{patternDisplay}</span>
+            <span className="pull-right">
```
Comment (edmundoa) [inline graylog2-web-interface/src/components/grok-patterns/GrokPatternInput.jsx]: I find the button placement a bit odd, I would try moving it to the header.

Before:
screen shot 2018-06-25 at 17 28 50

After:
screen shot 2018-06-25 at 17 28 56

File: graylog2-web-interface/src/components/grok-patterns/GrokPatternInput.jsx [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+          `${pattern.pattern.substring(0, maxLength)} ...` : pattern.pattern;
+        const active = index === this.state.activeListItem;
+        this.shownListItems.push(pattern.name);
+        return (
+          <ListGroupItem id={`list-item-${index}`}
+                         header={pattern.name}
+                         bsStyle={active ? 'info' : undefined}
+                         onKeyDown={this._onPatternFilterKeyDown}
+                         key={pattern.name}>
+            <span>{patternDisplay}</span>
```
Comment (edmundoa) [inline graylog2-web-interface/src/components/grok-patterns/GrokPatternInput.jsx]: I would use the full pattern in here, and then let CSS take care of hiding the content that is not visible with:

```
    text-overflow: ellipsis;
    white-space: nowrap;
    overflow: hidden;
    display: block;
```

With that all patterns will use all available space when rendering on the screen, and displaying ellipsis when the content is too large for that.

File: graylog2-web-interface/src/components/bootstrap/BootstrapModalForm.jsx [hunk 6]
```diff
@@ -21,6 +21,7 @@ class BootstrapModalForm extends React.Component {
     onModalClose: () => {},
     onSubmitForm: undefined,
     onCancel: () => {},
+    bsSize: 'large',
```
Comment (edmundoa) [inline graylog2-web-interface/src/components/bootstrap/BootstrapModalForm.jsx]: Most modals we use didn't set this property, so I guess we should leave this as `undefined` by default.

File: graylog2-web-interface/src/components/bootstrap/BootstrapModalWrapper.jsx [hunk 7]
```diff
@@ -11,7 +11,8 @@ class BootstrapModalWrapper extends React.Component {
     onOpen: () => {},
     onClose: () => {},
     onHide: () => {},
-  }
+    bsSize: 'large',
```
Comment (edmundoa) [inline graylog2-web-interface/src/components/bootstrap/BootstrapModalWrapper.jsx]: Same as above.

File: graylog2-web-interface/src/components/grok-patterns/GrokPatternInput.jsx [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+  _onPatternFilterKeyDown = (e) => {
+    const ARROW_DOWN = 40;
+    const ARROW_UP = 38;
+    const ENTER = 13;
+    const listItem = this.shownListItems[this.state.activeListItem];
+
+    let activeListItem = 0;
+    let domElement;
+    switch (e.keyCode) {
```
Comment (edmundoa) [inline graylog2-web-interface/src/components/grok-patterns/GrokPatternInput.jsx]: Moving down with the keyboard has issues and sometimes scrolls the outer component, in this case the modal:

scroll

**Comment by edmundoa** [comment]
It looks like the test snapshot needs to be updated:

```
[ERROR] FAIL src/components/users/UserPreferencesButton.test.jsx
[ERROR]   ● UserPreferencesButton › should load user data when user clicks edit button
[ERROR] 
[ERROR]     expect(value).toMatchSnapshot()
[ERROR]     
[ERROR]     Received value does not match stored snapshot 1.
[ERROR]     
[ERROR]     - Snapshot
[ERROR]     + Received
[ERROR]     
[ERROR]     @@ -21,11 +21,10 @@
[ERROR]           </Button>
[ERROR]           <UserPreferencesModal
[ERROR]             userName="Full"
[ERROR]           >
[ERROR]             <BootstrapModalForm
[ERROR]     -         bsSize="large"
[ERROR]               cancelButtonText="Cancel"
[ERROR]               formProps={Object {}}
[ERROR]               onCancel={[Function]}
[ERROR]               onModalClose={[Function]}
[ERROR]               onModalOpen={[Function]}
[ERROR]     @@ -33,22 +32,20 @@
[ERROR]               submitButtonDisabled={false}
[ERROR]               submitButtonText="Save"
[ERROR]               title="Preferences for user Full"
[ERROR]             >
[ERROR]               <BootstrapModalWrapper
[ERROR]     -           bsSize="large"
[ERROR]                 onClose={[Function]}
[ERROR]                 onHide={[Function]}
[ERROR]                 onOpen={[Function]}
[ERROR]                 showModal={false}
[ERROR]               >
[ERROR]                 <Modal
[ERROR]                   animation={true}
[ERROR]                   autoFocus={true}
[ERROR]                   backdrop={true}
[ERROR]                   bsClass="modal"
[ERROR]     -             bsSize="large"
[ERROR]                   dialogComponentClass={[Function]}
[ERROR]                   enforceFocus={true}
[ERROR]                   keyboard={true}
[ERROR]                   manager={
[ERROR]                     ModalManager {
[ERROR] 
[ERROR]       30 |     const instance = e().mount(<UserPreferencesButton userName={userName} />);
[ERROR]       31 | 
[ERROR]     > 32 |     expect(instance).toMatchSnapshot();
[ERROR]       33 |     expect(instance.find('button')).toBeDefined();
[ERROR]       34 | 
[ERROR]       35 |     instance.find('button').simulate('click');
[ERROR]       
[ERROR]       at Object.<anonymous> (src/components/users/UserPreferencesButton.test.jsx:32:22)
[ERROR] 
[ERROR]  › 1 snapshot test failed.
```

**Comment by edmundoa** [comment]
I have found another issue when editing a grok pattern and testing the edited pattern, but it seems to be related to code outside of this PR. Anyway, I'll describe the issue and (@kmerz) you can decide if you want to fix it here or not.

When editing a grok pattern, the contents of the pattern text area seem to be ignored if I test the edited pattern using the previous name. For example:
screen shot 2018-06-28 at 17 10 01

But if I change the name slightly, it works as expected:
screen shot 2018-06-28 at 17 10 10


I think the issue here is related to the way we register patterns, it looks like the stored patterns replace the one we want to test:
https://github.com/Graylog2/graylog2-server/blob/365d26fff714cf37b360d09a337b3138437f663a/graylog2-server/src/main/java/org/graylog2/grok/MongoDbGrokPatternService.java#L119-L122

**Pull request merged**